espnet2/asr_transducer/beam_search_transducer.py

```diff
@@ -0,0 +1,638 @@
+"""Search algorithms for Transducer models."""
+
+from dataclasses import dataclass
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+import numpy as np
+import torch
+
+from espnet2.asr_transducer.decoder.abs_decoder import AbsDecoder
+from espnet2.asr_transducer.joint_network import JointNetwork
+
+
+@dataclass
+class Hypothesis:
+    """Default hypothesis definition for Transducer search algorithms."""
+
+    score: float
+    yseq: List[int]
+    dec_state: Optional[Tuple[torch.Tensor, Optional[torch.Tensor]]]
+    lm_state: Optional[Union[Dict[str, Any], List[Any]]] = None
+
+
+@dataclass
+class ExtendedHypothesis(Hypothesis):
+    """Extended hypothesis definition for NSC beam search and mAES."""
+
+    dec_out: torch.Tensor = None
```

Comment:
It can't be None but because I'm extending `Hypothesis`, I could not set mandatory parameters after optional ones (`lm_state` from `Hypothesis` in that case).
I'm not sure what's the proper way here?